File: samples/profile/competitors/python/time/time_bench.py
#!/usr/bin/env python3
# Python time competitors — the same two workloads the Cajeta side runs:
# time-instant-arith (1M) and time-localdate-arith (100K), using stdlib datetime.
# One schema-conformant CSV row per benchmark. Cross-check = the exact closed-form
# sum (500006500000 / 5000050000).
import os
import resource
import sys
import time
from datetime import date, datetime, timedelta, timezone

N_INSTANT = 1000000
N_DATE = 100000
INSTANT_REF = 500006500000
DATE_REF = 5000050000
EPOCH_ORDINAL = date(1970, 1, 1).toordinal()


def env(k, d):
    v = os.environ.get(k)
    return v if v else d


def peak_rss_kb():
    return resource.getrusage(resource.RUSAGE_SELF).ru_maxrss


def stats(samples):
    s = sorted(samples)
    n = len(s)
    return s[0], s[n // 2], sum(s) // n, s[min(n - 1, n * 95 // 100)]


def emit(run_id, ts, bench, input_size, warmup, trials, st, ok):
    mn, med, mean, p95 = st
    mops = (input_size / med * 1e9 / 1e6) if med > 0 else 0.0
    pyver = env("PROFILE_LANG_VERSION", "")
    status = "ok" if ok else "invalid"
    return (
        f"1,{run_id},{ts},{bench},time,,{input_size},,{input_size},python,{pyver},datetime,cpython,-OO,"
        f"{warmup},{trials},{mn},{med},{mean},{p95},{mops:.2f},Mop/s,{peak_rss_kb()},"
        f"-1,-1,-1,-1,{status},{str(ok).lower()},,"
    )


def bench(warmup, trials, f, check):
    for _ in range(warmup):
        f()
    samples = []
    ok = True
    for _ in range(trials):
        t0 = time.perf_counter_ns()
        r = f()
        samples.append(time.perf_counter_ns() - t0)
        ok = check(r)
    return samples, ok


def instant_arith():
    utc = timezone.utc
    seven = timedelta(seconds=7)
    s = 0
    for i in range(N_INSTANT):
        t = datetime.fromtimestamp(i, utc)
        t2 = t + seven
        s += int(t2.timestamp())
    return s


def localdate_arith():
    one = timedelta(days=1)
    s = 0
    for i in range(N_DATE):
        d = date.fromordinal(EPOCH_ORDINAL + i)
        d2 = d + one
        s += d2.toordinal() - EPOCH_ORDINAL
    return s


def main():
    run_id = env("PROFILE_RUN_ID", "local")
    ts = env("PROFILE_RUN_TS", "")
    warmup = int(env("PROFILE_WARMUP", "3"))
    trials = int(env("PROFILE_TRIALS", "10"))
    out = []

    s, ok = bench(warmup, trials, instant_arith, lambda r: r == INSTANT_REF)
    out.append(emit(run_id, ts, "time-instant-arith", N_INSTANT, warmup, trials, stats(s), ok))

    s, ok = bench(warmup, trials, localdate_arith, lambda r: r == DATE_REF)
    out.append(emit(run_id, ts, "time-localdate-arith", N_DATE, warmup, trials, stats(s), ok))

    sys.stdout.write("\n".join(out) + "\n")


if __name__ == "__main__":
    main()

File: samples/profile/competitors/java/time/TimeBench.java
// Java time competitors — the same two workloads the Cajeta side runs:
// time-instant-arith (1M) and time-localdate-arith (100K), using the java.time
// package (the immutable value-object date/time API). One schema-conformant CSV
// row per benchmark, competitors/columns.txt order. Cross-check = the exact
// closed-form epoch sum (500006500000 / 5000050000), identical to the Go/Rust/
// C++/Python runners. All math in UTC (no DST), so days are exactly 86400 s and
// epoch days advance by one.
//
// java.time.Instant / java.time.LocalDate are immutable, so each step allocates a
// fresh value; we accumulate the epoch second / epoch day into a long sum and
// return it (consumed by the cross-check) so the JIT can't elide the loop.
//
// Notes: timings include JVM JIT warmup (honoring PROFILE_WARMUP/TRIALS like the
// other runners; the large-N inner loops are OSR-compiled well within a trial).
// alloc_bytes is per-thread bytes allocated for one execution, via the HotSpot
// com.sun.management.ThreadMXBean — the Java-native analog of Go's TotalAlloc
// delta / Rust's global allocator / C++ operator new override.
import com.sun.management.ThreadMXBean;
import java.lang.management.ManagementFactory;
import java.nio.file.Files;
import java.nio.file.Path;
import java.time.Instant;
import java.time.LocalDate;
import java.util.Arrays;
import java.util.function.LongSupplier;
import java.util.function.LongPredicate;

public class TimeBench {
    static final int N_INSTANT = 1000000;
    static final int N_DATE = 100000;
    static final long INSTANT_REF = 500006500000L;
    static final long DATE_REF = 5000050000L;
    static final ThreadMXBean TMX = (ThreadMXBean) ManagementFactory.getThreadMXBean();

    static String env(String k, String d) { String v = System.getenv(k); return (v == null || v.isEmpty()) ? d : v; }

    static long peakRssKb() {
        try {
            for (String line : Files.readAllLines(Path.of("/proc/self/status")))
                if (line.startsWith("VmHWM:")) return Long.parseLong(line.replaceAll("[^0-9]", ""));
        } catch (Exception e) { }
        return -1;
    }

    static long allocOf(Runnable r) {
        long id = Thread.currentThread().threadId();
        long b = TMX.getThreadAllocatedBytes(id);
        r.run();
        return TMX.getThreadAllocatedBytes(id) - b;
    }

    static void emit(String runId, String ts, String bench, int input, String lib,
                     int warmup, int trials, long[] samples, boolean ok, long alloc) {
        Arrays.sort(samples);
        int n = samples.length;
        long sum = 0; for (long x : samples) sum += x;
        int idx = Math.min(n - 1, n * 95 / 100);
        long mn = samples[0], med = samples[n / 2], mean = sum / n, p95 = samples[idx];
        double mops = med > 0 ? (double) input / (double) med * 1e9 / 1e6 : 0.0;
        String status = ok ? "ok" : "invalid";
        System.out.printf(
            "1,%s,%s,%s,time,,%d,,%d,java,%s,%s,stdlib,-server,%d,%d,%d,%d,%d,%d,%.2f,Mop/s,%d,-1,%d,-1,-1,%s,%b,,%n",
            runId, ts, bench, input, input, env("PROFILE_LANG_VERSION", ""), lib,
            warmup, trials, mn, med, mean, p95, mops, peakRssKb(), alloc, status, ok);
    }

    static void benchI64(String runId, String ts, String bench, int input, String lib,
                         int warmup, int trials, LongSupplier f, LongPredicate check) {
        for (int i = 0; i < warmup; i++) { long unused = f.getAsLong(); if (unused == Long.MIN_VALUE) System.out.print(""); }
        long[] samples = new long[trials];
        boolean ok = true;
        for (int i = 0; i < trials; i++) {
            long t0 = System.nanoTime();
            long r = f.getAsLong();
            samples[i] = System.nanoTime() - t0;
            ok = check.test(r);
        }
        long alloc = allocOf(() -> { long r = f.getAsLong(); if (r == Long.MIN_VALUE) System.out.print(""); });
        emit(runId, ts, bench, input, lib, warmup, trials, samples, ok, alloc);
    }

    public static void main(String[] args) {
        String runId = env("PROFILE_RUN_ID", "local");
        String ts = env("PROFILE_RUN_TS", "");
        int warmup = Integer.parseInt(env("PROFILE_WARMUP", "3"));
        int trials = Integer.parseInt(env("PROFILE_TRIALS", "10"));

        // instant-arith: Instant.ofEpochSecond(i).plusSeconds(7).getEpochSecond()
        benchI64(runId, ts, "time-instant-arith", N_INSTANT, "java.time", warmup, trials, () -> {
            long sum = 0;
            for (long i = 0; i < N_INSTANT; i++) {
                Instant t = Instant.ofEpochSecond(i);
                Instant t2 = t.plusSeconds(7);
                sum += t2.getEpochSecond();
            }
            return sum;
        }, r -> r == INSTANT_REF);

        // localdate-arith: epoch day i → +1 day → epoch day (UTC, exact 86400s/day)
        benchI64(runId, ts, "time-localdate-arith", N_DATE, "java.time", warmup, trials, () -> {
            long sum = 0;
            for (long i = 0; i < N_DATE; i++) {
                LocalDate t = LocalDate.ofEpochDay(i);
                LocalDate t2 = t.plusDays(1);
                sum += t2.toEpochDay();
            }
            return sum;
        }, r -> r == DATE_REF);
    }
}
